### ISSUE TITLE
ComfyUI LTX-2 Lora format loading not all keys in ComfyUI

### ISSUE THREAD

**Issue by kabachuha** [bghira/SimpleTuner#2349]
Edit: I seem to have figured it out, don't use the script yet!

---

When I'm trying to load LTX-2 loras in native ComfyUI with the LoRA format , I get a lot of the following errors:

<details>
lora key not loaded: diffusion_model.connectors.audio_connector.transformer_blocks.0.attn1.to_k.alpha
lora key not loaded: diffusion_model.connectors.audio_connector.transformer_blocks.0.attn1.to_k.lora_A.weight
lora key not loaded: diffusion_model.connectors.audio_connector.transformer_blocks.0.attn1.to_k.lora_B.weight
lora key not loaded: diffusion_model.connectors.audio_connector.transformer_blocks.0.attn1.to_out.0.alpha
lora key not loaded: diffusion_model.connectors.audio_connector.transformer_blocks.0.attn1.to_out.0.lora_A.weight
lora key not loaded: diffusion_model.connectors.audio_connector.transformer_blocks.0.attn1.to_out.0.lora_B.weight
lora key not loaded: diffusion_model.connectors.audio_connector.transformer_blocks.0.attn1.to_q.alpha
lora key not loaded: diffusion_model.connectors.audio_connector.transformer_blocks.0.attn1.to_q.lora_A.weight
lora key not loaded: diffusion_model.connectors.audio_connector.transformer_blocks.0.attn1.to_q.lora_B.weight
lora key not loaded: diffusion_model.connectors.audio_connector.transformer_blocks.0.attn1.to_v.alpha
lora key not loaded: diffusion_model.connectors.audio_connector.transformer_blocks.0.attn1.to_v.lora_A.weight
lora key not loaded: diffusion_model.connectors.audio_connector.transformer_blocks.0.attn1.to_v.lora_B.weight
lora key not loaded: diffusion_model.connectors.audio_connector.transformer_blocks.1.attn1.to_k.alpha
lora key not loaded: diffusion_model.connectors.audio_connector.transformer_blocks.1.attn1.to_k.lora_A.weight
lora key not loaded: diffusion_model.connectors.audio_connector.transformer_blocks.1.attn1.to_k.lora_B.weight
lora key not loaded: diffusion_model.connectors.audio_connector.transformer_blocks.1.attn1.to_out.0.alpha
lora key not loaded: diffusion_model.connectors.audio_connector.transformer_blocks.1.attn1.to_out.0.lora_A.weight
lora key not loaded: diffusion_model.connectors.audio_connector.transformer_blocks.1.attn1.to_out.0.lora_B.weight
lora key not loaded: diffusion_model.connectors.audio_connector.transformer_blocks.1.attn1.to_q.alpha
lora key not loaded: diffusion_model.connectors.audio_connector.transformer_blocks.1.attn1.to_q.lora_A.weight
lora key not loaded: diffusion_model.connectors.audio_connector.transformer_blocks.1.attn1.to_q.lora_B.weight
lora key not loaded: diffusion_model.connectors.audio_connector.transformer_blocks.1.attn1.to_v.alpha
lora key not loaded: diffusion_model.connectors.audio_connector.transformer_blocks.1.attn1.to_v.lora_A.weight
lora key not loaded: diffusion_model.connectors.audio_connector.transformer_blocks.1.attn1.to_v.lora_B.weight
lora key not loaded: diffusion_model.connectors.video_connector.transformer_blocks.0.attn1.to_k.alpha
lora key not loaded: diffusion_model.connectors.video_connector.transformer_blocks.0.attn1.to_k.lora_A.weight
lora key not loaded: diffusion_model.connectors.video_connector.transformer_blocks.0.attn1.to_k.lora_B.weight
</details>

It looks like there is a severe mismatch of names between Comfy and the model's format in this repository, because ComfyUI's LTX-2 model looks very different.

https://github.com/Comfy-Org/ComfyUI/blob/master/comfy/ldm/lightricks/av_model.py

https://github.com/Comfy-Org/ComfyUI/blob/master/comfy/ldm/lightricks/model.py

Meantime I did a conversion script for a trained LoRA (DeepSeek):

```python
import torch
from safetensors.torch import load_file, save_file
import argparse
import os
import re

def convert_lora_keys(lora_dict):
    """
    Convert LoRA keys from the original format to ComfyUI's format.
    
    Original format targets:
    - diffusion_model.connectors.audio_connector.transformer_blocks.N.attn1.to_*
    - diffusion_model.connectors.video_connector.transformer_blocks.N.attn1.to_*
    
    ComfyUI format expects:
    - diffusion_model.transformer_blocks.N.attn1.* (for video)
    - diffusion_model.transformer_blocks.N.audio_attn1.* (for audio)
    """
    
    converted_dict = {}
    
    for key, tensor in lora_dict.items():
        new_key = key
        
        # Check if this is an audio connector key
        if 'audio_connector.transformer_blocks.' in key:
            # Remove 'connectors.audio_connector.' and change attn1 to audio_attn1
            new_key = key.replace('connectors.audio_connector.', '')
            if '.attn1.' in new_key:
                new_key = new_key.replace('.attn1.', '.audio_attn1.')
        
        # Check if this is a video connector key  
        elif 'video_connector.transformer_blocks.' in key:
            # Remove 'connectors.video_connector.'
            new_key = key.replace('connectors.video_connector.', '')
        
        # Also handle any remaining connector references
        if 'connectors.transformer_blocks.' in new_key:
            new_key = new_key.replace('connectors.', '')
        
        converted_dict[new_key] = tensor
    
    return converted_dict

def verify_conversion(converted_dict):
    """
    Verify that the conversion was successful.
    """
    print("\nVerifying conversion...")
    
    # Check for remaining connector references
    connector_refs = [k for k in converted_dict.keys() if 'connector' in k or 'connectors' in k]
    
    if connector_refs:
        print(f"❌ Found {len(connector_refs)} keys still containing 'connector':")
        for key in connector_refs[:5]:
            print(f"  {key}")
        if len(connector_refs) > 5:
            print(f"  ... and {len(connector_refs) - 5} more")
        return False
    
    # Check expected patterns
    expected_patterns = [
        r'diffusion_model\.transformer_blocks\.\d+\.attn1\.',
        r'diffusion_model\.transformer_blocks\.\d+\.audio_attn1\.',
    ]
    
    all_valid = True
    for pattern in expected_patterns:
        matching_keys = [k for k in converted_dict.keys() if re.search(pattern, k)]
        if matching_keys:
            print(f"✓ Found {len(matching_keys)} keys matching pattern: {pattern}")
        else:
            print(f"⚠ No keys found matching pattern: {pattern}")
    
    # Count by type
    video_keys = [k for k in converted_dict.keys() if 'transformer_blocks.' in k and 'audio_attn1' not in k and 'attn1' in k]
    audio_keys = [k for k in converted_dict.keys() if 'audio_attn1' in k]
    
    print(f"\nSummary:")
    print(f"  Video attention keys: {len(video_keys)}")
    print(f"  Audio attention keys: {len(audio_keys)}")
    print(f"  Total keys: {len(converted_dict)}")
    
    # Show sample of converted keys
    print(f"\nSample converted keys:")
    sample_keys = []
    for pattern in expected_patterns:
        matches = [k for k in converted_dict.keys() if re.search(pattern, k)]
        if matches:
            sample_keys.extend(matches[:2])
    
    for i, key in enumerate(sample_keys[:10]):
        print(f"  {key}")
    
    return True

def main():
    parser = argparse.ArgumentParser(description='Convert LoRA keys for ComfyUI LTXAV model')
    parser.add_argument('input_path', type=str, help='Path to input safetensors file')
    parser.add_argument('output_path', type=str, help='Path to output safetensors file')
    parser.add_argument('--verify-only', action='store_true', help='Only verify without saving')
    
    args = parser.parse_args()
    
    # Load the original LoRA file
    print(f"Loading LoRA file from {args.input_path}")
    lora_dict = load_file(args.input_path)
    
    print(f"Original keys: {len(lora_dict)}")
    print("Sample original keys:")
    for i, key in enumerate(list(lora_dict.keys())[:5]):
        print(f"  {key}")
    
    # Convert the keys
    print("\nConverting keys...")
    converted_dict = convert_lora_keys(lora_dict)
    
    # Verify conversion
    success = verify_conversion(converted_dict)
    
    if not success:
        print("\n⚠ Conversion issues detected. Attempting alternative conversion...")
        
        # Try alternative approach for any remaining issues
        converted_dict = {}
        for key, tensor in lora_dict.items():
            new_key = key
            
            # More aggressive conversion
            if 'connectors' in new_key:
                # Remove everything between 'diffusion_model.' and 'transformer_blocks.'
                parts = new_key.split('.')
                try:
                    # Find indices
                    dm_idx = parts.index('diffusion_model')
                    tb_idx = parts.index('transformer_blocks')
                    
                    # Keep diffusion_model and transformer_blocks, remove everything in between
                    if tb_idx > dm_idx + 1:
                        parts = parts[:dm_idx + 1] + parts[tb_idx:]
                        new_key = '.'.join(parts)
                except ValueError:
                    pass
            
            # Convert audio attention
            if '.attn1.' in new_key and 'audio' in key.lower():
                new_key = new_key.replace('.attn1.', '.audio_attn1.')
            
            converted_dict[new_key] = tensor
        
        print("\nRetrying verification after alternative conversion...")
        success = verify_conversion(converted_dict)
    
    if not args.verify_only:
        if success:
            # Save the converted file
            print(f"\n✅ Conversion successful! Saving to {args.output_path}")
            save_file(converted_dict, args.output_path)
            print("Done!")
        else:
            print(f"\n⚠ Issues found during conversion. File may not work correctly.")
            save_choice = input("Save anyway? (y/n): ")
            if save_choice.lower() == 'y':
                save_file(converted_dict, args.output_path)
                print(f"Saved to {args.output_path}")
    else:
        print("\nVerification complete. Use without --verify-only flag to save.")

if __name__ == "__main__":
    main()
```

But I still didn't resolve

```python
lora key not loaded: text_embedding_projection.aggregate_embed.lora_A.weight
lora key not loaded: text_embedding_projection.aggregate_embed.lora_B.weight
```

I guess, before the fix, I'd stick to ["to_k", "to_q", "to_v", "to_out.0"]

**Comment by kabachuha** [comment]
Slightly offtopic, but the loss curves look... interesting

<img width="577" height="522" alt="Image" src="https://github.com/user-attachments/assets/d41823fd-539f-4e0e-9bb9-1a8b26253ca4" />

(Fully vanilla, no CREPA and stuff)

**Comment by bghira** [comment]
my loss looks nothing like that. it is in the 0.2-0.5 range

**Comment by bghira** [comment]
the comfyui lora_format is from the actual LTX-2 LoRA formats. i figured Comfy would use them. if not, oh well. fuck that project. 😄 they can introduce more conversions and handle simpletuner native format then.

**Comment by kabachuha** [comment]
> Slightly offtopic, but the loss curves look... interesting

Yeah, the result is all noise. I wonder what is the cause. Maybe the vid files were not read correctly after all? Or vae/encoder troubles. I'll check it

**Comment by bghira** [comment]
thanks. i'm looking into the issues detecting audio.

**Comment by bghira** [comment]
make sure you're not using some pre-quantised text encoder, i tried the unsloth gemma3 and got crazy high loss / NaN

**Comment by bghira** [comment]
also if you have like 4k or 1080p videos in the set i now realise aspect bucket worker count has to be reduced because 12 of these in memory can be like 50gb of *sysram 🤯

**Comment by kabachuha** [comment]
@bghira I fixed it (the loss). Shouldn't have set vae_dtype to fp16 manually. Now it's 0.3-0.5 initially.

**Comment by bghira** [comment]
ok, i guess that should to into the LTXVideo2 class method for checking user settings and raise an exception

**Comment by kabachuha** [comment]
I seems to have figured out this. By expecting the state dict of a SimpleTuner-made LoRA, I see that the rest of the keys are normal.

<img width="640" height="380" alt="Image" src="https://github.com/user-attachments/assets/850e1f70-892a-4d23-af92-17549e40caa0" />

It's just the "connectors" (and text_embedding_projections) the LoRA is also applied to.

The obvious solution is to either set default peft targets, so they include all the vanilla-compatible attentions, or to simply ban the connectors from q/k/v loras being applied to (hardcoded exclusion list)

**Comment by bghira** [comment]
do you have good validation output? on AMD, mine are very bad 😓

**Comment by kabachuha** [comment]
I think my results are decent so far (3H of training)

Here is a sample on a random object in ComfyUI

https://github.com/user-attachments/assets/74013df5-ac1a-403f-ae64-6780c7a0e682

https://github.com/user-attachments/assets/a977a5fe-1200-4842-9319-4e978fb4a3b3

The checkpoint will be ready in 20 minutes, I'll upload it to Huggingface for test + the configs and dataset link

I'm on Nvidia, so can't say thing about AMD :(

**Comment by bghira** [comment]
i mean using the trainers' built-in validation

**Comment by kabachuha** [comment]
I don't use it, actually. For me it's more convenient to launch the checkpoints in Comfy from time to time, when I see the loss doing interesting things.

**Comment by bghira** [comment]
ah, if you run in server mode, you can just send an api call to start validation manually and keep the schedule far enough apart that it never runs automated

**Comment by kabachuha** [comment]
Okay, it doesn't generalize on very surreal concepts yet. I'll think I'll put it on the night with 3x longer, higher LR and TREAD and then consider to upload, because it's undercooked.

https://github.com/user-attachments/assets/94a3dbee-3d18-41e5-b5bf-8aad8bcfc7c6

As for this issue about the state dict, there need to be some exclusion rule for the "connector" blocks (and text_embedding_projection). Or to form the comprehensive list of PEFT *inclusion*

If we print the state dictionary of LoRA with

```python
from safetensors import safe_open
f = safe_open("/media/kabachuha/xiangliu/simpletunerx/ltx2-models-crush-new/checkpoint-1296/pytorch_lora_weights.safetensors", framework="pt", device=0)
f.keys()
```

Then I can see these patterns (for to_q, to_k, to_v, to_out.0):

```
connectors.audio_connector.transformer_blocks.0.attn1.to_k
...
connectors.video_connector.transformer_blocks.0.attn1.to_k
...

transformer_blocks.0.attn1.to_k
...
transformer_blocks.0.audio_attn1.to_k
...
transformer_blocks.0.audio_to_video_attn.to_k
...
transformer_blocks.0.video_to_audio_attn.to_k
```

The first two need to be eliminated

From the PEFT docs https://huggingface.co/docs/peft/main/en/package_reference/lora#peft.LoraConfig

> target_modules (Optional[Union[List[str], str]]) — The names of the modules to apply the adapter to. If this is specified, only the modules with the specified names will be replaced. When passing a string, a regex match will be performed. When passing a list of strings, either an exact match will be performed or it is checked if the name of the module ends with any of the passed strings. If this is specified as ‘all-linear’, then all linear/Conv1D modules are chosen (if the model is a PreTrainedModel, the output layer excluded). If this is not specified, modules will be chosen according to the model architecture. If the architecture is not known, an error will be raised — in this case, you should specify the target modules manually. To avoid targeting any modules (because you want to apply target_parameters), set target_modules=[].

It seems, a regex expression has to be passed there instead of a list to exclude the connectors from training.

**Comment by bghira** [comment]
https://github.com/huggingface/peft/issues/2044

**Comment by kabachuha** [comment]
Ok, found, thank you!

**Comment by kabachuha** [comment]
It would fit alongside `peft_lora_target_modules` in the settings / model defaults

**Comment by kabachuha** [comment]
I set `exclude_modules=".*connector.*|.*embedding.*",` in `simpletuner/helpers/models/common.py`, let's see how it'll go on

**Comment by kabachuha** [comment]
Now the SD starts fine:

```
(comfy) kabachuha@kabachuha-mipt:~/ComfyUI$ cat ks.txt
diffusion_model.transformer_blocks.0.attn1.to_k.alpha
diffusion_model.transformer_blocks.0.attn1.to_k.lora_A.weight
diffusion_model.transformer_blocks.0.attn1.to_k.lora_B.weight
diffusion_model.transformer_blocks.0.attn1.to_out.0.alpha
diffusion_model.transformer_blocks.0.attn1.to_out.0.lora_A.weight
...
```

And ComfyUI has no warnings.

---

Technically, I'm satisfied now and the issue is solved

**Comment by bghira** [comment]
if you'd like to push a PR, you could set DEFAULT_LORA_EXCLUDE_TARGETS to None on the base foundation and then override it for LTX2 and use that as the source of this value

**Comment by kabachuha** [comment]
Good idea. As for the PR, it's quite late where I am, so I can only do it in some hours. If you can do before that, I'd appreciate

**Comment by bghira** [comment]
im stuck trying to figure out what's going on with the validation quality here all night probably
